modules/default/weather/providers/yr.js

```diff
@@ -0,0 +1,622 @@
+						clearInterval(attemptFetchWeather);
+						this.getWeatherDataFromYrOrCache(resolve, reject);
+					}
+				}, 100);
+			} else {
+				this.getWeatherDataFromYrOrCache(resolve, reject);
+			}
+		});
+	},
+
+	getWeatherDataFromYrOrCache(resolve, reject) {
+		localStorage.setItem("yrIsFetchingWeatherData", "true");
+
+		let weatherData = this.getWeatherDataFromCache();
+		if (this.weatherDataIsValid(weatherData)) {
+			localStorage.removeItem("yrIsFetchingWeatherData");
+			Log.debug("Weather data found in cache.");
+			resolve(weatherData);
+		} else {
+			this.getWeatherDataFromYr(weatherData?.downloadedAt)
+				.then((weatherData) => {
+					Log.debug("Got weather data from yr.");
+					if (weatherData) {
+						this.cacheWeatherData(weatherData);
+					} else {
+						//Undefined if unchanged
+						weatherData = this.getWeatherDataFromCache();
+					}
+					resolve(weatherData);
+				})
+				.catch((err) => {
+					Log.error(err);
+					reject("Unable to get weather data from Yr.");
+				})
+				.finally(() => {
+					localStorage.removeItem("yrIsFetchingWeatherData");
+				});
+		}
+	},
+
+	weatherDataIsValid(weatherData) {
+		return (
+			weatherData &&
+			weatherData.timeout &&
+			0 < moment(weatherData.timeout).diff(moment()) &&
+			(!weatherData.geometry || !weatherData.geometry.coordinates || !weatherData.geometry.coordinates.length < 2 || (weatherData.geometry.coordinates[0] === this.config.lat && weatherData.geometry.coordinates[1] === this.config.lon))
+		);
+	},
+
+	getWeatherDataFromCache() {
+		const weatherData = localStorage.getItem("weatherData");
+		if (weatherData) {
+			return JSON.parse(weatherData);
+		} else {
+			return undefined;
+		}
+	},
+
+	getWeatherDataFromYr(currentDataFetchedAt) {
+		const requestHeaders = [{ name: "Accept", value: "application/json" }];
+		if (currentDataFetchedAt) {
+			requestHeaders.push({ name: "If-Modified-Since", value: currentDataFetchedAt });
+		}
+
+		const expectedResponseHeaders = ["expires", "date"];
+
+		return this.fetchData(this.getForecastUrl(), "json", requestHeaders, expectedResponseHeaders)
+			.then((data) => {
+				if (!data || !data.headers) return data;
+				data.timeout = data.headers.find((header) => header.name === "expires").value;
+				data.downloadedAt = data.headers.find((header) => header.name === "date").value;
+				data.headers = undefined;
+				return data;
+			})
+			.catch((err) => {
+				Log.error("Could not load weather data.", err);
+				throw new Error(err);
+			});
+	},
+
+	getForecastUrl() {
+		if (!this.config.lat) {
+			Log.error("Latitude not provided.");
+			throw new Error("Latitude not provided.");
+		}
+		if (!this.config.lon) {
+			Log.error("Longitude not provided.");
+			throw new Error("Longitude not provided.");
+		}
+
+		const lat = this.config.lat.toString();
+		const lon = this.config.lon.toString();
+		const altitude = this.config.altitude ?? 0;
+
+		if (lat.includes(".") && lat.split(".")[1].length > 4) {
```

Comment:
I think it would be more user friendly if the logic would gracefully cut the decimals to max 4 and just issue a warning instead of not starting at all